### ISSUE TITLE
When using email templates, missing_sessions contains duplicate IP's 

### ISSUE THREAD

**Issue by mngan** [respawner/peering-manager#221]
### Environment
* Python version:  3.6.9
* Peering Manager version:  master

### Steps to Reproduce
If you have multiple routers registered in PeeringDB for a given IX, the query to generate the missing IP's returns duplicates of the peer IP's for the missing sessions. 

The handling of multiple routers connected to the same IX also has other odd side effects, such as this template reporting on missing sessions multiple time for the IXes that have multiple routers. There isn't an easy way to fix this with the way things are organized. 

This is the template code that I used to generate the email
```
{%- for entry in internet_exchanges | sort(attribute="internet_exchange.name") %}
  {%- with %}
  {%- set sessions = entry.sessions | selectattr("enabled") | rejectattr("bgp_state", "equalto", "established") | list %}
  {%- set missing = entry.missing_sessions.values() | sum(start=[]) %}
  {%- set router_ipv4 = entry.internet_exchange.ipv4_address %}
  {%- set router_ipv6 = entry.internet_exchange.ipv6_address %}
  {%- if sessions | length > 0 or missing | length > 0 %}
{{"| %-114s |" % entry.internet_exchange.name}}
{{"+"+"-"*116+"+"}}
{{"| %-39s | %-39s | %-12s | %-6s | %-6s |" % ("Your IP Address", "Valve IP Address", "BGP State", "Rcvd", "Advt") }}
{{"+"+"-"*116+"+"}}
    {%- for session in sessions | sort(attribute="ip_address.version,ip_address")%}
      {%- with router_ip = router_ipv4 if session.ip_address.version == 4 else router_ipv6 %}
{{"| %-39s | %-39s | %12s | %6d | %6d |" % (session.ip_address, router_ip, session.bgp_state, session.received_prefix_count, session.advertised_prefix_count) }}
      {%- endwith %}
    {%- endfor %}
    {%- for family, missing_ips in entry.missing_sessions | dictsort() %}
      {%- for ip_address in missing_ips | sort %}
        {%- with router_ip = router_ipv4 if ip_address.version == 4 else router_ipv6 %}
{{"| %-39s | %-39s | %12s | %6s | %6s |" % (ip_address, router_ip, "missing", "", "") }}
        {%- endwith %}
      {%- endfor %}
    {%- endfor %}
{{"+"+"-"*116+"+"}}
  {%- endif %}
  {%- endwith %}
{%- endfor %}
```

### Expected Behavior
I would expect the missing_sessions field to only have a single entry for each missing session

### Observed Behavior

```
+--------------------------------------------------------------------------------------------------------------------+
| DE-CIX Madrid: DE-CIX Madrid Peering LAN                                                                           |
+--------------------------------------------------------------------------------------------------------------------+
| Your IP Address                         | Valve IP Address                        | BGP State    | Rcvd   | Advt   |
+--------------------------------------------------------------------------------------------------------------------+
| 185.1.68.87                             | 185.1.68.86                             |      missing |        |        |
| 185.1.68.87                             | 185.1.68.86                             |      missing |        |        |
| 2001:7f8:a0::2a:0:1                     | 2001:7f8:a0::7f4e:0:2                   |      missing |        |        |
| 2001:7f8:a0::2a:0:1                     | 2001:7f8:a0::7f4e:0:2                   |      missing |        |        |
+--------------------------------------------------------------------------------------------------------------------+
```

**Comment by mngan** [comment]
I am not sure this actually fixes it. I can still see multiple entries turn up when I look at AS42 in the test data I sent you. I think the problem is actually rooted in get_ix_networks_for_asn. This will return a network_ixlan object for each connection to the IX, and the code will evaluate all the connections in find_potential_ix_peering_sessions leading to duplicates. Perhaps if the potential_ix_peering_sessions variable were a set rather than a list, this would solve it in this case. The only other place where this is called is in get_common_internet_exchanges, which fortunately treats the results more like a set, so this isn't an issue.

**Comment by respawner** [comment]
I've been trying to reproduce this issue and found that it only appears when you have several connections to the same IX. Can you confirm this behavior?

If so my guess is that you are correct on the possible points of failure.

**Comment by mngan** [comment]
Yes, this is correct. It only occurs when you have multiple connections to the same IX

**Comment by respawner** [comment]
There is a quick and dirty solution (IMHO) to patch that issue

```diff
diff --git a/peering/models.py b/peering/models.py
index 9f2d68f..d6134a9 100644
--- a/peering/models.py
+++ b/peering/models.py
@@ -242,10 +242,12 @@ class AutonomousSystem(ChangeLoggedModel, TaggableModel, TemplateModel):
             missing_sessions = {"ipv6": [], "ipv4": []}
             for session in self.potential_internet_exchange_peering_sessions:
                 for prefix in internet_exchange.get_prefixes():
-                    if session in ipaddress.ip_network(prefix):
-                        missing_sessions["ipv{}".format(session.version)].append(
-                            session
-                        )
+                    if (
+                        session not in missing_sessions["ipv6"]
+                        and session not in missing_sessions["ipv4"]
+                    ):
+                        if session in ipaddress.ip_network(prefix):
+                            missing_sessions[f"ipv{session.version}"].append(session)

             ix_and_sessions.append(
                 {
```

However I think that there is a more refactoring of the way we keep track of connection to an IX today. Currently if you have 2 or more connections to the same IX you have to create 2 or more IX instances in Peering Manager. It might be interesting to introduce a `Connection` or `Port` or whatever object that would change the game. Instead of having multiple IX instance, you'll have one with multiple `Connection` on it (a connection being, IP addresses to use, as well as other details).

What I can propose is to in the first place patch this issue in the "dirty" way and then work a the new way to model connections on an IX. This work could include the fix for #216 as well.

**Comment by jlixfeld** [comment]
+1 on refactoring this.  This will affect my implementation shortly as well, as I'm planning on introducing multiple links to the same IX in multiple markets.

**Comment by respawner** [comment]
Close this issue as the final fix will be discussed in #246.
